ContentMigrationBatch#12WD (#7587)
GoogleChrome/web.dev: src/site/content/en/blog/stick-to-compositor-only-properties-and-manage-layer-count/index.md

<figure>
{% Img src="image/T4FyVKpzu4WKF1kBNvXepbi08t52/4to9zPKpICVNFCPl7qwV.jpg", alt="The pixel pipeline with no layout or paint.", width="800", height="122" %}
</figure>

In order to achieve this you will need to stick to changing properties that can be handled by the compositor alone. Today there are only two properties for which that is true - `transform`s and `opacity`:

<figure>
{% Img src="image/T4FyVKpzu4WKF1kBNvXepbi08t52/VoLzvR2akAnQfKFVyyM9.jpg", alt="The properties you can animate without triggering layout or paint.", width="800", height="600" %}
</figure>

The caveat for the use of `transform`s and `opacity` is that the element on which you change these properties should be on _its own compositor layer_. In order to make a layer you must promote the element, which we will cover next.

Note: If you’re concerned that you may not be able to limit your animations to just those properties, take a look at the [FLIP principle](https://aerotwist.com/blog/flip-your-animations), which may help you remap animations to changes in transforms and opacity from more expensive properties.

## Promote elements that you plan to animate

> As we mentioned in the “[Simplify paint complexity and reduce paint areas](https://developers.google.com/web/fundamentals/performance/rendering/simplify-paint-complexity-and-reduce-paint-areas)” section, you should promote elements that you plan to animate (within reason, don’t overdo it!) to their own layer:

Reviewer: ```suggestion
As we mentioned in the "[Simplify paint complexity and reduce paint areas](https://developers.google.com/web/fundamentals/performance/rendering/simplify-paint-complexity-and-reduce-paint-areas)" section, you should promote elements that you plan to animate (within reason, don’t overdo it!) to their own layer:
```

```css
.moving-element {
  will-change: transform;
}
```

> Or, for older browsers, or those that don’t support will-change:

Reviewer: ```suggestion
Or, for older browsers, or those that don't support will-change:
```

Reviewer: linting error might be this smart quote. Try committing this.

Reviewer: or the other one above

<figure>
{% Img src="image/T4FyVKpzu4WKF1kBNvXepbi08t52/VoLzvR2akAnQfKFVyyM9.jpg", alt="The properties you can animate without triggering layout or paint.", width="800", height="600" %}
</figure>

The caveat for the use of `transform`s and `opacity` is that the element on which you change these properties should be on _its own compositor layer_. In order to make a layer you must promote the element, which we will cover next.

Note: If you’re concerned that you may not be able to limit your animations to just those properties, take a look at the [FLIP principle](https://aerotwist.com/blog/flip-your-animations), which may help you remap animations to changes in transforms and opacity from more expensive properties.

## Promote elements that you plan to animate

As we mentioned in the “[Simplify paint complexity and reduce paint areas](https://developers.google.com/web/fundamentals/performance/rendering/simplify-paint-complexity-and-reduce-paint-areas)” section, you should promote elements that you plan to animate (within reason, don’t overdo it!) to their own layer:

```css
.moving-element {

>     will-change: transform;

Reviewer 2: ```suggestion
  will-change: transform;
```

}
```

Or, for older browsers, or those that don’t support will-change:

```css
.moving-element {

>     transform: translateZ(0);

Reviewer 2: ```suggestion
  transform: translateZ(0);
```

}

> ```
> 
> 
> This gives the browser the forewarning that changes are incoming and, depending on what you plan to change, the browser can potentially make provisions, such as creating compositor layers.

Reviewer 2: ```suggestion
```

This gives the browser the forewarning that changes are incoming and, depending on what you plan to change, the browser can potentially make provisions, such as creating compositor layers.
```

## Manage layers and avoid layer explosions

It’s perhaps tempting, then, knowing that layers often help performance, to promote all the elements on your page with something like the following:

```css

> * {
>     will-change: transform;
>     transform: translateZ(0);
> }

Reviewer 2: ```suggestion
* {
  will-change: transform;
  transform: translateZ(0);
}
```

```

Which is a roundabout way of saying that you’d like to promote every single element on the page. The problem here is that every layer you create requires memory and management, and that’s not free. In fact, on devices with limited memory the impact on performance can far outweigh any benefit of creating a layer. Every layer’s textures needs to be uploaded to the GPU, so there are further constraints in terms of bandwidth between CPU and GPU, and memory available for textures on the GPU.

Warning: Do not promote elements unnecessarily.

## Use Chrome DevTools to understand the layers in your app

>   <figure>
>     {% Img src="image/T4FyVKpzu4WKF1kBNvXepbi08t52/QNAAH0VJDRjtu9iC6KVE.jpg", alt="The toggle for the paint profiler in Chrome DevTools.", width="278", height="129" %}
>   </figure>

Reviewer 2: ```suggestion
<figure>
  {% Img src="image/T4FyVKpzu4WKF1kBNvXepbi08t52/QNAAH0VJDRjtu9iC6KVE.jpg", alt="The toggle for the paint profiler in Chrome DevTools.", width="278", height="129" %}
</figure>
```
